### ISSUE TITLE
dd.read_hdf fails with pandas >= 0.24.0

### ISSUE THREAD

**Issue by st-bender** [dask/dask#5934]
Hi
`dd.read_hdf()` fails to read some files which are perfectly fine to be opened with `pandas` (and with `dask` + `pandas < 0.24.0`). I believe that this is due to a rather recent change in `pandas` pandas-dev/pandas#21543 (issue pandas-dev/pandas#21372) in an attempt to speed up hdf5 reading in some cases.

The consequence is that the `.keys()` function returns an empty list on some files (pandas-dev/pandas#29916). The `.keys()` method is used by `dd.read_hdf()` to get the list of keys in the file to read, see:
https://github.com/dask/dask/blob/54deb7fa4cdbbf3e124be6834224272be27ace65/dask/dataframe/io/hdf.py#L298-L302

I am not sure where the correct place is to fix this (`dask`, `pandas`, or `pytables`?). One way would be to use the pytables handle itself `_handle.walk_nodes()` instead of relying on pandas' `.keys()` in the lines above (the `dask` fix). However, that could hit the performance again in some cases, possibly related to #1353, #1408 
OTOH, I'd take a slow but working version over a non-working version any day.

Cheers.

**Comment by quasiben** [comment]
@st-bender thanks for linking to the Pandas issues. I am not sure dask should do anything here just yet.  It seems like the discussion in https://github.com/pandas-dev/pandas/issues/29916 is moving towards a PR the pandas devs would be willing to accept which dask could then leverage

**Comment by st-bender** [comment]
Hi @quasiben,
Thanks for taking it on. I am not sure *if* let alone *when* it will get fixed in `pandas`.
However, given that I can open the files and access the datasets just fine with pandas, but **not** with dask, one should at least consider a more permanent solution within dask.
Cheers.

**Comment by st-bender** [comment]
Hi @quasiben and devs, after some discussion around the pandas issue and the pull requests, the proposed solution is to add a keyword to the `.keys()` method. This keyword would have to be used by dask then, rendering it incompatible with any pandas version < 1.x (without the keyword).

The alternative would be to solve that within dask by replacing the `.keys()` call with the underlying pytables `walk_nodes`, which seems to have been there for a while and compatible with older versions. The patch would be a rather straight forward port of the pandas pull requests  pandas-dev/pandas#32401, pandas-dev/pandas#32723.

Or rewrite the hdf5 reading completely (#1353, #1408).

Any suggestions?

**Comment by quasiben** [comment]
Thanks @st-bender for continuing here.  Dask already used the `keys` method, correct ?  If dask does nothing here and the PR is merged, users would have to upgrade pandas to the next release for the fix, correct ?

**Comment by st-bender** [comment]
> Thanks @st-bender for continuing here. Dask already used the `keys` method, correct ?

Yes, see my issue report above for the code snippet.

> If dask does nothing here and the PR is merged, users would have to upgrade pandas to the next release for the fix, correct ?

Even with the proposed fix in pandas, dask will need to change the code in question to use the extra keyword. Users will need to upgrade both dask and pandas, and newer dask versions will no longer work with older pandas versions.

Dask doing nothing will not fix this issue.

**Comment by st-bender** [comment]
HI there, is there any activity planned?
I hope everybody is fine these times.
It looks like the issue is a bit stuck on the pandas side because it seems to be difficult to decide how to best/least break the API. And even then will the fix not be backwards compatible.

It seems that dask is about the only project that needs that fix, and it is getting a bit annoying to hand-patch dask after every update. The fix itself is very simple, replace `hdf.keys()` in line 299 above by the list comprehension:
```python
[n._v_pathname for n in hdf._handle.walk_nodes("/", classname="Table")]
```
(adapted from https://github.com/pandas-dev/pandas/pull/32723)

Cheers.

**Comment by TomAugspurger** [comment]
I'd be a bit uncomfortable relying on private attributes and methods like that.

**Comment by st-bender** [comment]
> I'd be a bit uncomfortable relying on private attributes and methods like that.

Understandable, but it is relying on the public API that introduced this bug in the first place, because that changed. The `._handle.` interface to the underlying pytables structure is at least backwards compatible.
OTOH, this issue **might** get fixed on the pandas side after all, but it seems to be a bit stuck in the pipeline, and it would require dask to check the pandas version because of the added keyword:
https://github.com/pandas-dev/pandas/pull/32723#issuecomment-601385565 and
https://github.com/pandas-dev/pandas/pull/32723#issuecomment-601396775 (both almost 2 months old now).

**Comment by TomAugspurger** [comment]
Perhaps you could help push those along by reviewing them.

**Comment by st-bender** [comment]
[CC'd from #6204]
I found some time to play a little bit around, basically pandas-dev/pandas#21543 broke compatibility by skipping non-pandas objects. I tried to adapt the approach from pandas to dask and wrote a test case for a strangely mixed case. Here is a solution I came up with, still depends on `hdf._handle` and things; patch to hdf.py and a test:
https://gist.github.com/st-bender/97d78da1cc6a3ef1b917e02455522dd6

Still not sure what goes wrong with py36 and pandas 0.23.4 though, but maybe the approach fixes that?
